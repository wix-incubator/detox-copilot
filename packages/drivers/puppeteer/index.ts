import {
  TestingFrameworkAPICatalog,
  TestingFrameworkDriver,
} from "@wix-pilot/core";
import * as puppeteer from "puppeteer-core";
import WebTestingFrameworkDriverHelper from "@wix-pilot/web-utils";
import fs from "fs";

export class PuppeteerFrameworkDriver implements TestingFrameworkDriver {
  private executablePath?: string;
  private driverUtils: WebTestingFrameworkDriverHelper;

  constructor(executablePath?: string) {
    this.setCurrentPage = this.setCurrentPage.bind(this);
    this.getCurrentPage = this.getCurrentPage.bind(this);
    this.executablePath = executablePath;
    this.driverUtils = new WebTestingFrameworkDriverHelper();
  }

  /**
   * Gets the current page identifier
   */
  getCurrentPage(): puppeteer.Page | undefined {
    return this.driverUtils.getCurrentPage() as puppeteer.Page | undefined;
  }

  /**
   * Sets the current page identifier, must be set if the driver needs to interact with a specific page
   */
  setCurrentPage(page: puppeteer.Page): void {
    this.driverUtils.setCurrentPage(page);
  }

  /**
   * Injects bundeled code to page and marks important elements in the DOM
   */
  async injectCodeAndMarkElements(page: puppeteer.Page): Promise<void> {
    const isInjected = await page.evaluate(
      () => typeof window.driverUtils?.markImportantElements === "function",
    );

    if (!isInjected) {
      await page.addScriptTag({
        content: fs.readFileSync(this.driverUtils.getBundledCodePath(), "utf8"),
      });
      console.log("Bundled script injected into the page.");
    } else {
      console.log("Bundled script already injected. Skipping injection.");
    }

    await page.evaluate(() => window.driverUtils.markImportantElements());
  }

  /**
   * Mark the elements and separates them to categories
   */
  async manipulateStyles(page: puppeteer.Page): Promise<void> {
    await page.evaluate(() => {
      window.driverUtils.manipulateElementStyles();
    });
  }

  /**
   * Clean up page style changes
   */
  async cleanUpStyleChanges(page: puppeteer.Page): Promise<void> {
    await page.evaluate(() => {
      window.driverUtils.cleanupStyleChanges();
    });
  }

  /**
   * @inheritdoc
   */
  async captureSnapshotImage(): Promise<string | undefined> {
    return await this.driverUtils.captureSnapshotImage();
  }

  /**
   * @inheritdoc
   */
  async captureViewHierarchyString(): Promise<string> {
    return await this.driverUtils.captureViewHierarchyString();
  }

  /**
   * @inheritdoc
   */
  get apiCatalog(): TestingFrameworkAPICatalog {
    return {
      name: "Puppeteer",
      description:
        "Puppeteer is a Node library which provides a high-level API to control Chrome or Chromium over the DevTools Protocol.\nYou can assume that puppeteer is already imported (as `puppeteer`).",
      context: {
        getCurrentPage: this.getCurrentPage,
        setCurrentPage: this.setCurrentPage,
        puppeteer,
      },
      categories: [
        {
          title: "Browser",
          items: [
            {
              signature: "const browser = await puppeteer.launch([options])",
              description: "Launches a new browser instance.",
              example: `const browser = await puppeteer.launch({\`headless: "false"\`, executablePath: "${this.executablePath}" });`,
              guidelines: [
                `Executable path is required always, use the path: ${this.executablePath}`,
                "Options can specify `headless`, `slowMo`, `args`, etc.",
                "Useful for running tests in a headless browser environment.",
                'Prefer passing `headless: "new"` to `puppeteer.launch() unless mentioned that ' +
                  "it is required not to (e.g. launching with GUI was mentioned).",
              ],
            },
            {
              signature: "await browser.close()",
              description: "Closes the browser instance.",
              example: "await getCurrentPage().browser().close();",
              guidelines: [
                "Allows to close the browser after finishing a test flow.",
                "Useful for cleaning up resources and freeing memory.",
              ],
            },
            {
              signature: "getCurrentPage().setUserAgent(userAgent)",
              description: "Overrides the default user agent string.",
              example: 'getCurrentPage().setUserAgent("UA-TEST");',
              guidelines: [
                "Affects the value of `navigator.userAgent`.",
                "Useful for simulating different browsers or bots.",
              ],
            },
          ],
        },
        {
          title: "Current page management",
          items: [
            {
              signature: "const page = getCurrentPage()",
              description:
                "Gets the current page instance. Can return `undefined` if no page is set.",
              example: "const page = getCurrentPage();",
            },
            {
              signature: "await setCurrentPage(page)",
              description:
                "Sets the current page instance for the driver to interact with (required if setting a new page).",
              example: "await setCurrentPage(page);",
            },
          ],
        },
        {
          title: "Matchers",
          items: [
            {
              signature:
                'document.querySelector(\'[aria-pilot-category="categoryName"][aria-pilot-index="index"]\')',
              description:
                "Selects a specific element within a category based on its index.",
              example: `const firstButton = await page.evaluate(() => document.querySelector('[aria-pilot-category="button"][aria-pilot-index="27"]');`,
              guidelines: [
                "Replace `categoryName` with the desired category and `index` with the specific index as a string.",
                "Indexing is zero-based and increments per category as elements are found.",
                "Use this to interact with or verify a specific instance of a category, ensuring the exact element is targeted.",
              ],
            },
          ],
        },
      ],
    };
  }
}
