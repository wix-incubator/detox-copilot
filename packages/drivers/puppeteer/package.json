{
    "name": "@pilot/puppeterr-driver",
    "version": "0.0.1",
    "description": "Puppeterr driver for pilot and copilot usage",
    "main": "dist/index.js",
    "types": "dist/index.d.ts",
    "files": [
      "dist"
    ],
    "publishConfig": {
      "registry": "https://registry.npmjs.org"
    },
    "repository": {
      "type": "git",
      "url": "git+https://github.com/wix-incubator/detox-copilot.git"
    },
    "scripts": {
      "build": "tsc && tsc-alias",
      "test": "echo No tests avialable for this package",
      "test:example": "jest",
      "bump-version": "npm version patch && git push && git push --tags",
      "release": "npm run test && npm run bump-version && npm run build && npm publish",
      "lint": "eslint . --ext .ts",
      "lint:fix": "eslint . --ext .ts --fix",
      "type-check": "tsc --noEmit"
    },
    "bugs": {
      "url": "https://github.com/wix-incubator/detox-copilot/issues"
    },

    "dependencies": {
     "@pilot/core": "^0.0.37",
     "@pilot/drivers-utils": "^0.0.1"
    },

    "devDependencies": {
      "puppeteer": "^20.8.0",
      "axios": "^1.7.9"

    }
  }
