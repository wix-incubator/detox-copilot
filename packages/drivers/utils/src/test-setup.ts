import * as path from 'path';
import { build } from 'esbuild';

export async function bundleDriverUtils(): Promise<string> {
    try {
        const result = await build({
            entryPoints: [path.resolve(__dirname, './index.ts')],
            bundle: true,
            write: false,
            format: 'iife',
            globalName: 'driverUtils',
            platform: 'browser',
            target: 'es2015',
            minify: false,
            outfile: 'out.js',
        });

        if (!('outputFiles' in result) || !result.outputFiles?.[0]?.text) {
            throw new Error('No output generated from esbuild');
        }

        // Add a wrapper to handle default export
        const bundled = `
            (function(global) {
                ${result.outputFiles[0].text}
                // Handle ES module default export
                if (typeof driverUtils !== 'undefined' && driverUtils.default) {
                    global.driverUtils = driverUtils.default;
                }
            })(typeof window !== 'undefined' ? window : global);
        `;

        return bundled;
    } catch (error) {
        console.error('Bundling failed:', error);
        throw error;
    }
}
