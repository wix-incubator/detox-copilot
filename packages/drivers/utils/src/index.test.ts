import * as puppeteer from "puppeteer";
import { toMatchImageSnapshot } from "jest-image-snapshot";
import { bundleUtils } from "./test-setup";

expect.extend({ toMatchImageSnapshot });

declare global {
  interface Window {
    driverUtils: typeof import("./index").default;
  }
}

describe("driver-utils Puppeteer Tests", () => {
  let browser: puppeteer.Browser;
  let page: puppeteer.Page;
  let bundledCode: string;

  jest.setTimeout(30000);

  beforeAll(async () => {
    try {
      bundledCode = await bundleUtils();
      browser = await puppeteer.launch({
        devtools: false,
        args: ["--no-sandbox", "--disable-setuid-sandbox"],
      });

      page = await browser.newPage();
      page.on("console", (msg) => console.log("Browser log:", msg.text()));
      page.on("pageerror", (err) => console.error("Page error:", err));
      page.on("error", (err) => console.error("Browser error:", err));
      page.setDefaultNavigationTimeout(10000);

      await page.goto(`file://${__dirname}/test-pages/test-page.html`);
      await page.addScriptTag({ content: bundledCode });
    } catch (error) {
      console.error("Setup failed:", error);
      throw error;
    }
  });

  afterAll(async () => {
    if (browser) {
      await browser.close();
    }
  });

  describe("Element Marking Functionality", () => {
    beforeEach(async () => {
      // Reset state before each test
      await page.evaluate(() => {
        window.driverUtils.cleanupStyleChanges();
      });
    });

    it("should mark all supported elements with correct categories", async () => {
      const categories = await page.evaluate(() => {
        window.driverUtils.markImportantElements();
        const elements = document.querySelectorAll("[aria-pilot-category]");
        return Array.from(elements).map((el) => ({
          tag: el.tagName.toLowerCase(),
          category: el.getAttribute("aria-pilot-category"),
          index: el.getAttribute("aria-pilot-index"),
        }));
      });

      expect(categories).toMatchSnapshot("element-categories");
    });

    it("should handle hidden elements correctly", async () => {
      const hiddenElementsResult = await page.evaluate(() => {
        // Add some hidden elements
        const div = document.createElement("div");
        div.innerHTML = `
                    <button style="display: none">Hidden Button</button>
                    <a href="#" style="visibility: hidden">Hidden Link</a>
                    <input type="text" hidden>
                `;
        document.body.appendChild(div);

        // Mark elements without including hidden
        window.driverUtils.markImportantElements();

        // Check if hidden elements were marked
        const hiddenElements = Array.from(div.children);
        return hiddenElements.map((el) => ({
          tag: el.tagName.toLowerCase(),
          hasCategory: el.hasAttribute("aria-pilot-category"),
        }));
      });

      expect(hiddenElementsResult).toMatchSnapshot("hidden-elements-handling");
    });

    it("should mark hidden elements when includeHidden option is true", async () => {
      const hiddenElementsResult = await page.evaluate(() => {
        window.driverUtils.markImportantElements({ includeHidden: true });
        const elements = document.querySelectorAll("[aria-pilot-category]");
        return Array.from(elements).map((el) => ({
          tag: el.tagName.toLowerCase(),
          category: el.getAttribute("aria-pilot-category"),
          isHidden:
            window.getComputedStyle(el).display === "none" ||
            window.getComputedStyle(el).visibility === "hidden" ||
            el.hasAttribute("hidden"),
        }));
      });

      expect(hiddenElementsResult).toMatchSnapshot("hidden-elements-included");
    });
  });

  describe("Style Manipulation", () => {
    it("should apply correct styles to marked elements", async () => {
      await page.evaluate(() => {
        window.driverUtils.markImportantElements();
        window.driverUtils.manipulateElementStyles();
      });

      const styles = await page.evaluate(() => {
        const elements = document.querySelectorAll("[aria-pilot-category]");
        return Array.from(elements).map((el) => ({
          category: el.getAttribute("aria-pilot-category"),
          computedStyle: {
            border: window.getComputedStyle(el).border,
            position: window.getComputedStyle(el).position,
            margin: window.getComputedStyle(el).margin,
          },
        }));
      });

      expect(styles).toMatchSnapshot("applied-styles");
    });

    it("should properly clean up all applied styles", async () => {
      const stylesAfterCleanup = await page.evaluate(() => {
        window.driverUtils.cleanupStyleChanges();
        const styleElement = document.getElementById("aria-pilot-styles");
        const markedElements = document.querySelectorAll(
          "[aria-pilot-category]",
        );

        return {
          styleElementExists: styleElement !== null,
          elementsWithComputedStyles: Array.from(markedElements).map((el) => ({
            tag: el.tagName.toLowerCase(),
            border: window.getComputedStyle(el).border,
            position: window.getComputedStyle(el).position,
          })),
        };
      });

      expect(stylesAfterCleanup).toMatchSnapshot("styles-after-cleanup");
    });
  });

  describe("View Structure Extraction", () => {
    it("should generate clean DOM structure with only relevant elements", async () => {
      const structure = await page.evaluate(() => {
        window.driverUtils.markImportantElements();
        return window.driverUtils.extractCleanViewStructure();
      });

      expect(structure).toMatchSnapshot("clean-dom-structure");
    });

    it("should preserve allowed attributes and remove others", async () => {
      const attributesCheck = await page.evaluate(() => {
        // Add some elements with various attributes
        const div = document.createElement("div");
        div.innerHTML = `
                    <button class="btn" style="color: red" data-test="value" onclick="alert()">Test</button>
                    <a href="#link" target="_blank" rel="noopener" style="color: blue">Link</a>
                    <img src="test.jpg" alt="Test" width="100" height="100">
                `;
        document.body.appendChild(div);

        window.driverUtils.markImportantElements();
        const structure = window.driverUtils.extractCleanViewStructure();
        document.body.removeChild(div);

        return structure;
      });

      expect(attributesCheck).toMatchSnapshot("attribute-filtering");
    });
  });

  describe("Visual Regression Testing", () => {
    it("should match baseline snapshot with marked elements", async () => {
      await page.evaluate(() => {
        window.driverUtils.markImportantElements();
        window.driverUtils.manipulateElementStyles();
      });

      const image = await page.screenshot({ fullPage: true });
      expect(image).toMatchImageSnapshot({
        failureThreshold: 0.01,
        failureThresholdType: "percent",
      });
    });

    it("should match baseline snapshot after cleanup", async () => {
      await page.evaluate(() => {
        window.driverUtils.cleanupStyleChanges();
      });

      const image = await page.screenshot({ fullPage: true });
      expect(image).toMatchImageSnapshot({
        failureThreshold: 0.01,
        failureThresholdType: "percent",
      });
    });
  });

  describe.skip("Layout Preservation", () => {
    it("should not affect element positioning or spacing", async () => {
      const getMetrics = async () =>
        await page.evaluate(() => {
          const elements = document.querySelectorAll(
            "button, a, input, nav, ul, li",
          );
          return Array.from(elements).map((el) => {
            const rect = el.getBoundingClientRect();
            const style = window.getComputedStyle(el);
            return {
              tag: el.tagName.toLowerCase(),
              rect: {
                top: rect.top,
                left: rect.left,
                width: rect.width,
                height: rect.height,
              },
              margin: style.margin,
              padding: style.padding,
            };
          });
        });

      // Capture original layout metrics
      const originalMetrics = getMetrics();

      // Apply marking and styling
      await page.evaluate(() => {
        window.driverUtils.markImportantElements();
        window.driverUtils.manipulateElementStyles();
      });

      // Compare with new metrics
      const newMetrics = await getMetrics();

      expect(newMetrics).toEqual(originalMetrics);
    });
  });

  describe.skip("Clean View Structure", () => {
    it("should preserve element hierarchy and relationships", async () => {
      const structure = await page.evaluate(() => {
        window.driverUtils.markImportantElements();
        return window.driverUtils.extractCleanViewStructure();
      });

      // Verify nav contains ul contains li
      expect(structure).toMatch(/<nav[^>]*>\s*<ul[^>]*>\s*<li[^>]*>/);

      // Verify proper nesting and closing tags
      expect(structure).toMatch(/<nav[^>]*>.*<\/nav>/s);
      expect(structure).toMatch(/<ul[^>]*>.*<\/ul>/s);
      expect(structure).toMatch(/<li[^>]*>.*<\/li>/s);

      // Verify essential attributes
      expect(structure).toMatch(/data-category="semantic"/);
      expect(structure).toMatch(/data-category="list"/);
      expect(structure).toMatch(/href="#test"/);
      expect(structure).toMatch(/type="text"/);
    });

    it("should handle deeply nested structures", async () => {
      await page.evaluate(() => {
        // Add complex nested structure
        const complex = document.createElement("div");
        complex.innerHTML = `
                <nav data-testid="main-nav">
                    <ul>
                        <li>
                            <a href="#1">Link 1</a>
                            <ul>
                                <li><a href="#1.1">Sublink 1.1</a></li>
                                <li><a href="#1.2">Sublink 1.2</a></li>
                            </ul>
                        </li>
                    </ul>
                </nav>
            `;
        document.body.appendChild(complex);
      });

      const structure = await page.evaluate(() => {
        window.driverUtils.markImportantElements();
        return window.driverUtils.extractCleanViewStructure();
      });

      expect(structure).toMatch(/nav.*ul.*li.*a.*ul.*li.*a/s);
      expect(structure).toMatch(/href="#1\.1"/);
      expect(structure).toMatch(/href="#1\.2"/);
    });
  });
});
