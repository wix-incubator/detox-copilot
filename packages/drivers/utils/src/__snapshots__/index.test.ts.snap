// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`driver-utils Puppeteer Tests Element Marking Functionality should handle hidden elements correctly: hidden-elements-handling 1`] = `
[
  {
    "hasCategory": false,
    "tag": "button",
  },
  {
    "hasCategory": false,
    "tag": "a",
  },
  {
    "hasCategory": false,
    "tag": "input",
  },
]
`;

exports[`driver-utils Puppeteer Tests Element Marking Functionality should mark all supported elements with correct categories: element-categories 1`] = `
[
  {
    "category": "button",
    "index": "0",
    "tag": "button",
  },
  {
    "category": "link",
    "index": "0",
    "tag": "a",
  },
  {
    "category": "input",
    "index": "0",
    "tag": "input",
  },
  {
    "category": "table",
    "index": "0",
    "tag": "table",
  },
  {
    "category": "semantic",
    "index": "0",
    "tag": "nav",
  },
  {
    "category": "list",
    "index": "0",
    "tag": "ul",
  },
]
`;

exports[`driver-utils Puppeteer Tests Element Marking Functionality should mark hidden elements when includeHidden option is true: hidden-elements-included 1`] = `
[
  {
    "category": "button",
    "isHidden": false,
    "tag": "button",
  },
  {
    "category": "link",
    "isHidden": false,
    "tag": "a",
  },
  {
    "category": "input",
    "isHidden": false,
    "tag": "input",
  },
  {
    "category": "table",
    "isHidden": false,
    "tag": "table",
  },
  {
    "category": "semantic",
    "isHidden": false,
    "tag": "nav",
  },
  {
    "category": "list",
    "isHidden": false,
    "tag": "ul",
  },
  {
    "category": "button",
    "isHidden": true,
    "tag": "button",
  },
  {
    "category": "link",
    "isHidden": true,
    "tag": "a",
  },
  {
    "category": "input",
    "isHidden": true,
    "tag": "input",
  },
]
`;

exports[`driver-utils Puppeteer Tests Style Manipulation should apply correct styles to marked elements: applied-styles 1`] = `
[
  {
    "category": "button",
    "computedStyle": {
      "border": "2px outset rgb(0, 0, 0)",
      "margin": "0px",
      "position": "relative",
    },
  },
  {
    "category": "link",
    "computedStyle": {
      "border": "0px none rgb(0, 0, 238)",
      "margin": "0px",
      "position": "relative",
    },
  },
  {
    "category": "input",
    "computedStyle": {
      "border": "2px inset rgb(118, 118, 118)",
      "margin": "0px",
      "position": "relative",
    },
  },
  {
    "category": "table",
    "computedStyle": {
      "border": "0px none rgb(128, 128, 128)",
      "margin": "0px",
      "position": "relative",
    },
  },
  {
    "category": "semantic",
    "computedStyle": {
      "border": "0px none rgb(0, 0, 0)",
      "margin": "0px",
      "position": "relative",
    },
  },
  {
    "category": "list",
    "computedStyle": {
      "border": "0px none rgb(0, 0, 0)",
      "margin": "16px 0px",
      "position": "relative",
    },
  },
  {
    "category": "button",
    "computedStyle": {
      "border": "2px outset rgb(0, 0, 0)",
      "margin": "0px",
      "position": "relative",
    },
  },
  {
    "category": "link",
    "computedStyle": {
      "border": "0px none rgb(0, 0, 238)",
      "margin": "0px",
      "position": "relative",
    },
  },
  {
    "category": "input",
    "computedStyle": {
      "border": "2px inset rgb(118, 118, 118)",
      "margin": "0px",
      "position": "relative",
    },
  },
]
`;

exports[`driver-utils Puppeteer Tests Style Manipulation should properly clean up all applied styles: styles-after-cleanup 1`] = `
{
  "elementsWithComputedStyles": [
    {
      "border": "2px outset rgb(0, 0, 0)",
      "position": "static",
      "tag": "button",
    },
    {
      "border": "0px none rgb(0, 0, 238)",
      "position": "static",
      "tag": "a",
    },
    {
      "border": "2px inset rgb(118, 118, 118)",
      "position": "static",
      "tag": "input",
    },
    {
      "border": "0px none rgb(128, 128, 128)",
      "position": "static",
      "tag": "table",
    },
    {
      "border": "0px none rgb(0, 0, 0)",
      "position": "static",
      "tag": "nav",
    },
    {
      "border": "0px none rgb(0, 0, 0)",
      "position": "static",
      "tag": "ul",
    },
    {
      "border": "2px outset rgb(0, 0, 0)",
      "position": "static",
      "tag": "button",
    },
    {
      "border": "0px none rgb(0, 0, 238)",
      "position": "static",
      "tag": "a",
    },
    {
      "border": "2px inset rgb(118, 118, 118)",
      "position": "static",
      "tag": "input",
    },
  ],
  "styleElementExists": false,
}
`;

exports[`driver-utils Puppeteer Tests View Structure Extraction should generate clean DOM structure with only relevant elements: clean-dom-structure 1`] = `
"<html>
  <head>
  </head>
  <body>
  </body>
</html>
"
`;

exports[`driver-utils Puppeteer Tests View Structure Extraction should preserve allowed attributes and remove others: attribute-filtering 1`] = `
"<html>
  <head>
  </head>
  <body>
  </body>
</html>
"
`;
